akka-stream/src/main/scala/akka/stream/impl/Throttle.scala

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import akka.stream.ThrottleMode.{ EnforcingMode, ShapingMode }
+import akka.stream.impl.fusing.GraphStages.SimpleLinearGraphStage
+import akka.stream.stage._
+import akka.stream._
+
+import scala.concurrent.duration.{ FiniteDuration, _ }
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+private[stream] trait ThrottleTimeCalculationHelper {
+  def now(): Long = System.currentTimeMillis()
+
+  def calculateSpeed(counter: Double, time: FiniteDuration): Long =
+    ((counter / time.toMillis) * 1024 * 1024).toLong
+
+  def scale(e: Int): Long = e.toLong << 20
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] abstract class AbstractThrottle[T](cost: Int,
+                                                   per: FiniteDuration,
+                                                   maximumBurst: Int,
+                                                   mode: ThrottleMode)
+  extends SimpleLinearGraphStage[T] with ThrottleTimeCalculationHelper {
+
+  def calculateElementCost(elem: T): Long
+  val timerName: String
+
+  val enforcedModeException = new IllegalStateException("Maximum throttle throughput exceeded")
```

Comment:
this messes up the stacktrace. No need to cache this exception. Also, I would prefer a domain-specific exception for this: RateExceededException.